### DISCUSSION TITLE
Testing `getStaticProps` with Jest

### DISCUSSION THREAD

**Discussion by phidah** [vercel/next.js#15365] [Help]
I have a page that retrieves data from Prismic through `getStaticProps`:

```javascript
export const getStaticProps = async ({ req }) => {
  const currentFeature = (await Client(req).query(
    [
      Prismic.Predicates.at('document.type', 'feature'),
      Prismic.Predicates.at('my.feature.featured', true)
    ],
    {
      orderings: '[my.billy_feature.order]'
    }
  )).results[0]

  return {
    props: {
      currentFeature
    }
  }
}

export default class Page extends React.Component {
  render () {
    const { currentFeature } = this.props

    return (
      <div>
        <h2>{RichText.asText(currentFeature.data.name)}</h2>
        <RichText render={currentFeature.data.body} />
      </div>
    )
  }
}

```

I am now trying to write a Jest test that tests that this works as intended:

```javascript
import React from 'react'
import renderer from 'react-test-renderer'
import Funktioner from '../pages/funktioner/index'

it('renders the features page unchanged', () => {
  const tree = renderer.create(<Funktioner />).toJSON()
  expect(tree).toMatchSnapshot()
})

```

Unfortunately the test fails:

```
 FAIL  tests/funktioner.test.js
  ✕ renders the features page unchanged (43 ms)

  ● renders the features page unchanged

    TypeError: Cannot read property 'data' of undefined

      47 |   render () {
      48 |     const { currentFeature } = this.props
    > 49 |     const thisOrder = currentFeature.data.order
```

I started debugging and it turns out the `getStaticProps` function was never called. I assume this is because I imported the `Page` class directly, thus bypassing Next.

The question is - what do I do if I want to test static generation like this?

**Comment by jamesmosier** [comment]
The best way to test the functionality of an entire page is via an end to end test, not a snapshot test. Your assumption is correct that if you just import the page directly your test will have no clue about `getStaticProps`.

Something like [cypress](https://www.cypress.io/) or [selenium](https://help.crossbrowsertesting.com/selenium-testing/getting-started/javascript/) would be good choices for E2E tests.

**Reply by toomuchdesign** [reply]
I recently tried to fill this testing gap by replicating some of the glue provided by Next.js in [this library I published as `next-page-tester`](https://github.com/toomuchdesign/next-page-tester).

Give a Next route, the library is supposed to take care of:

- resolving provided routes into the matching page component
- calling Next.js data fetching methods (`getServerSideProps` or `getStaticProps`) if the case
- set up a mocked next/router provider initialized with the expected values (to test `useRouter` and `withRouter`)
- instantiating the page component with the expected props

I'm still not sure whether this testing approach can make sense for Next.js users, but I'd be happy to hear back from anybody who decides to tried it out.

**Reply by khousuylong** [reply]
I have tried the 'next page tester', but encounter the problem 'The error below may be caused by using the wrong test environment, see https://jestjs.io/docs/en/configuration#testenvironment-string. Consider using the "jsdom" test environment.'

**Reply by mkrtchian** [reply]
@khousuylong [this issue](https://github.com/toomuchdesign/next-page-tester/issues/118) may help you.

**Comment by 173Cs** [comment]
Hey, change the class based component to functional component cause, getStaticProps will only work in a functional component
